### PR TITLE
Basic fucntionalities

### PR DESCRIPTION
**Pull request by SheheryarAli-Arbisoft**
These are the basic functionalities of the application.

### CODE REVIEW

File: front-end/src/components/Button/props.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import PropTypes from 'prop-types';
+
+export const propTypes = {
+  children: PropTypes.node,
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Button/props.js:4]: Re-arrange the proptypes. For example (strings together)

File: front-end/src/components/Button/props.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  variant: PropTypes.string,
+  fullWidth: PropTypes.bool,
+  size: PropTypes.string,
+  color: PropTypes.string,
+  type: PropTypes.string,
+};
+
+export const defaultProps = {
+  variant: 'contained',
+  fullWidth: false,
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Button/props.js]: Re-arrange order according to above.

File: front-end/src/components/Card/Card.js [hunk 3]
```diff
@@ -1,23 +1,90 @@
-import React from 'react';
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Card.js:1]: Organize imports: (Libraries, Redux Files, App components, UI components, Styled Components, Utils, Images)

File: front-end/src/components/Card/Card.js [hunk 4]
```diff
@@ -1,23 +1,90 @@
...
+import React, { useState, useEffect } from 'react';
+import moment from 'moment';
+import { WatchLater, ThumbUpAlt, ThumbUpAltOutlined } from '@material-ui/icons';
+import { useDispatch, useSelector } from 'react-redux';
+import { getData } from '../../redux/nodes/authentication/selectors';
+import { likePost, unlikePost } from '../../redux/nodes/entities/posts/actions';
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Card.js:8]: Also re-arrange actions together and selector together

File: front-end/src/components/Card/Card.js [hunk 5]
```diff
@@ -1,23 +1,90 @@
...
 import { Text } from '../Text';
+import { Button } from '../Button';
+import { CreateCommentForm } from '../Forms';
+
+export const Card = ({ post, ...rest }) => {
+  const dispatch = useDispatch();
+  const { userId } = useSelector(getData);
+  const { comments } = useSelector(getComments);
+  const { users } = useSelector(getUsers);
+  const [liked, setLiked] = useState(false);
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Card.js]: state name should be likesCount

File: front-end/src/components/Card/Card.js [hunk 6]
```diff
@@ -1,23 +1,90 @@
...
-        </Text>
+        <GridContainer>
+          <GridItem xs={12}>
+            <Text variant='h5' style={{ fontWeight: 'bold' }} gutterBottom>
+              {users[post.user].name}
+            </Text>
+          </GridItem>
+
+          <GridItem xs={12}>
+            <Text
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Card.js:45]: What text it is? Please rename it. Frome Example Title, Description

File: front-end/src/components/Card/Card.js [hunk 7]
```diff
@@ -1,23 +1,90 @@
...
+              gutterBottom
+              style={{ display: 'flex', alignItems: 'center', gap: '5px' }}
+            >
+              <WatchLater style={{ fontSize: 'inherit' }} />
+              {`${moment(post.timestamp).fromNow()}`}
+            </Text>
+          </GridItem>
+
+          <GridItem xs={12}>
+            <Text variant='body1' gutterBottom>
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Card.js:57]: Same as above.

File: front-end/src/components/Card/Card.js [hunk 8]
```diff
@@ -1,23 +1,90 @@
...
+            <Text variant='h6' gutterBottom>
+              Comments
+            </Text>
+          </GridItem>
+
+          <GridItem xs={12}>
+            <CreateCommentForm postId={post._id} />
+          </GridItem>
+
+          {post.comments.map(commentId => (
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Card.js]: Install lodash library and use map from that library.

File: front-end/src/components/Card/Comment.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import { GridContainer, GridItem } from '../Grid';
+import { Text } from '../Text';
+import { CreateReplyForm } from '../Forms';
+
+export const Comment = ({ commentId, comments, users }) => {
+  const [showInput, setShowInput] = useState(false);
+
+  return (
+    <GridItem xs={12}>
+      <div
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Comment.js]: Give styles with styled components

File: front-end/src/components/Card/Comment.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        }}
+      >
+        <GridContainer>
+          <GridItem xs={12}>
+            <Text variant='body1' style={{ fontWeight: 'bold' }}>
+              {users[comments[commentId].user].name}
+            </Text>
+          </GridItem>
+          <GridItem xs={12}>
+            <Text
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Comment.js]: Give styles with styled components

File: front-end/src/components/Card/Comment.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+          <GridItem xs={12}>
+            <Text variant='body1' gutterBottom>
+              {comments[commentId].content}
+            </Text>
+          </GridItem>
+          {!showInput && (
+            <GridItem xs={12}>
+              <Text
+                variant='body2'
+                color='textSecondary'
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Comment.js:54]: define all colors in the theme object and pass color using that.

File: front-end/src/components/Card/Reply.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import moment from 'moment';
+import { WatchLater } from '@material-ui/icons';
+import { GridContainer, GridItem } from '../Grid';
+import { Text } from '../Text';
+
+export const Reply = ({ replyId, comments, users }) => {
+  return (
+    <GridItem xs={12}>
+      <div
+        style={{
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Card/Reply.js]: Remove styles wherever you used and in elements and use styled-components to styles

File: front-end/src/components/Container/MainContainer/MainContainer.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import { StyledContainer } from './styled';
+import { propTypes, defaultProps } from './props';
+
+export const MainContainer = ({ children, ...rest }) => {
+  return <StyledContainer {...rest}>{children}</StyledContainer>;
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Container/MainContainer/MainContainer.js:6]: check spacing on both sides of {}

File: front-end/src/components/Input/props.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import PropTypes from 'prop-types';
+
+export const propTypes = {
```
Comment (BilalMaqsoodA) [inline front-end/src/components/Input/props.js:3]: Re-arrange order

File: front-end/src/pages/Login.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
+import React, { useEffect } from 'react';
```
Comment (BilalMaqsoodA) [inline front-end/src/pages/Login.js:1]: Organize imports: (Libraries, Redux Files, App components, UI components, Styled Components, Utils, Images)

**Review by BilalMaqsoodA — CHANGES_REQUESTED** [review]
We need a lot of refactoring. kindly, do these changes.

File: front-end/src/components/Input/props.js (hunk 14 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Input/props.js:3]: The order has been fixed.

File: front-end/src/components/Button/props.js (hunk 1 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Button/props.js:4]: The proptypes are now grouped together by type.

File: front-end/src/components/Button/props.js (hunk 2 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Button/props.js]: The order has been fixed.

File: front-end/src/components/Card/Card.js (hunk 5 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Card.js]: Liked is actually a flag to check if the post has been liked. This state has been renamed to postLiked.

File: front-end/src/components/Container/MainContainer/MainContainer.js (hunk 13 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Container/MainContainer/MainContainer.js:6]: This spacing is being managed by prettier. On save it formats according to this style.

File: front-end/src/pages/Login.js (hunk 15 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/pages/Login.js:1]: Import order has been fixed.

File: front-end/src/components/Card/Card.js (hunk 4 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Card.js:8]: The get data selector is actually coming from the authentication state. It has been renamed to getAuthData for better clarity.

File: front-end/src/components/Card/Card.js (hunk 3 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Card.js:1]: The import order has been fixed.

File: front-end/src/components/Card/Card.js (hunk 8 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Card.js]: Now using map function from lodash library.

File: front-end/src/components/Card/Comment.js (hunk 11 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Comment.js:54]: The color is actually a prop which is already defined in the default theme provided by Material UI. Material UI has been used as the component UI library.

File: front-end/src/components/Card/Card.js (hunk 6 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Card.js:45]: The type of text is controlled using the variant prop. On each value to variant props different format of text is displayed.

File: front-end/src/components/Card/Card.js (hunk 7 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Card.js:57]: Same response as above.

File: front-end/src/components/Card/Comment.js (hunk 9 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Comment.js]: Added styles with styled components.

File: front-end/src/components/Card/Comment.js (hunk 10 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Comment.js]: Same answer as above.

File: front-end/src/components/Card/Reply.js (hunk 12 above)
Comment (SheheryarAli-Arbisoft) [reply front-end/src/components/Card/Reply.js]: Replaced all style props with styled components and component props where needed.

**Pull request merged**